File: openclaw_safe/channels/whatsapp.py
"""WhatsApp channel adapter (Meta Cloud API)."""

from __future__ import annotations

from typing import Any

from .base import BaseChannel


class WhatsAppChannel(BaseChannel):
    """Adapter for the WhatsApp Business Cloud API."""

    channel_name = "whatsapp"

    def __init__(self, access_token: str, phone_number_id: str) -> None:
        self._access_token = access_token
        self._phone_number_id = phone_number_id

    def send_message(self, recipient_id: str, text: str, **kwargs: Any) -> None:
        # In production: POST to graph.facebook.com/v18.0/{phone-id}/messages
        raise NotImplementedError(
            "WhatsAppChannel.send_message requires a live access token and HTTP client."
        )

    def receive_message(self, payload: dict[str, Any]) -> dict[str, Any]:
        entry = payload.get("entry", [{}])[0]
        changes = entry.get("changes", [{}])[0]
        value = changes.get("value", {})
        messages = value.get("messages", [{}])
        msg = messages[0] if messages else {}
        return {
            "sender_id": msg.get("from", ""),
            "text": msg.get("text", {}).get("body", ""),
            "channel": self.channel_name,
            "raw": payload,
        }

File: src/channels/whatsapp.ts
import { BaseChannel, NormalisedMessage } from "./base";

export class WhatsAppChannel extends BaseChannel {
  readonly channelName = "whatsapp";

  constructor(private readonly accessToken: string, private readonly phoneNumberId: string) {
    super();
  }

  async sendMessage(_recipientId: string, _text: string): Promise<void> {
    throw new Error("WhatsAppChannel.sendMessage requires a live access token and HTTP client.");
  }

  receiveMessage(payload: unknown): NormalisedMessage {
    const p = payload as Record<string, unknown>;
    const entries = (p["entry"] as unknown[]) ?? [{}];
    const changes = ((entries[0] as Record<string, unknown>)["changes"] as unknown[]) ?? [{}];
    const value = (changes[0] as Record<string, unknown>)["value"] as Record<string, unknown> ?? {};
    const messages = (value["messages"] as unknown[]) ?? [{}];
    const msg = (messages[0] as Record<string, unknown>) ?? {};
    const textObj = (msg["text"] as Record<string, unknown>) ?? {};
    return {
      senderId: String(msg["from"] ?? ""),
      text: String(textObj["body"] ?? ""),
      channel: this.channelName,
      raw: payload,
    };
  }
}
